### PR TITLE
feat!: Update to support AWS provider v3.75 and newer (including v4.x)

### LINKED ISSUE #137: Support terraform-provider-aws 4.x
## Is your request related to a new offering from AWS?
No

## Is your request related to a problem? Please describe.
https://github.com/hashicorp/terraform-provider-aws 4.0.0 has breaking changes for the S3 bucket resource, see https://github.com/hashicorp/terraform-provider-aws/releases/tag/v4.0.0

## Describe the solution you'd like.
This module should work with https://github.com/hashicorp/terraform-provider-aws 4.x.x

## Describe alternatives you've considered.
We could abandon use of this module, but really don't want to ;)

## Additional context
N/A

### PR DESCRIPTION
**Pull request by antonbabenko**
Breaking changes:
- Yes, see file `UPGRADE-3.0.md`
- Requires AWS provider v3.75 or newer (including v4.x)

Fixes #137 
Closes #138

### CODE REVIEW

**Comment by gruckion** [comment]
Is there a way for me to use this now? Rather than waiting for it to be merged? How is that done in the Terraform world. Also 👍  for making the changes.

**Comment by rblaine95** [comment]
> Is there a way for me to use this now? Rather than waiting for it to be merged? How is that done in the Terraform world. Also 👍 for making the changes.

It's **_definitely_** not recommended to use in anything other than testing, and is probably very highly discouraged, but you can do this:

```hcl
module "s3" {
  source = "git::https://github.com/terraform-aws-modules/terraform-aws-s3-bucket.git?ref=fix-tf-provider-v4
}
```

**Comment by dirnikk** [comment]
Thanks Anton,
I think you missed some var declaration, but overall it's work for me as expected.

**Comment by rnikoopour** [comment]
HI All,

What needs to happen to get this across the finish line?  I'd be happy to contribute where I can.

This module has been a great help for me and I'd like for this get completed and merged.

**Comment by antonbabenko** [comment]
Hello everyone!

I have been recovering and resumed working on this PR earlier today after some thinking about potential resolutions mentioned here (https://github.com/hashicorp/terraform-provider-aws/issues/23106). One of the things I have been thinking about was how to make this module compatible with v3 AND v4 to minimize state migration for resources.

Anyway, I plan to finish this PR and have support for AWS provider v4 as soon as possible - Tuesday, 29th of March.

**Comment by ImperialXT** [comment]
in regards to state migration using moved blocks would be an easy win? no need for manual state moves by anyone? https://www.terraform.io/language/modules/develop/refactoring

File: versions.tf [hunk 1]
```diff
@@ -4,7 +4,7 @@ terraform {
   required_providers {
     aws = {
       source  = "hashicorp/aws"
-      version = "~> 3.69"
+      version = ">= 4.2.0"
```
Comment (vutny) [inline versions.tf:7]:
```suggestion
      version = ">= 3.75.0"
```
Hi @antonbabenko 
Thanks for your work on this.
Actually, Hashicorp "backported" those breaking changes for S3 resources into AWS provider version `3.75.0`.
So I think this is a clear way to keep the code with `3.x` release unless all major modules will support `4.x` in full.

Comment (antonbabenko) [reply versions.tf:7]: Yes, I will incorporate this change also. Thank you!

**Comment by antonbabenko** [comment]
This PR is included in version 3.0.0 :tada:

**Pull request merged**